### ISSUE TITLE
How to launch a Window from the Application class in WinUI?

### ISSUE THREAD

**Issue by akash07k** [microsoft/microsoft-ui-xaml#6725]
I'm creating a WinUI application which will run in background initially without launching a window.
Means I've removed the below lines from the OnLaunched method:
                    m_window = new MainWindow();
                    m_window.Activate();
Now the thing is that I've implemented a keyboard hook by which I want to give a shortcut key to the user to launch the window.
As soon as user is pressing the shortcut key, we are getting the exception such as: 
```
An exception of type 'System.Runtime.InteropServices.COMException' occurred in WinRT.Runtime.dll but was not handled in user code
The application called an interface that was marshalled for a different thread. (0x8001010E (RPC_E_WRONG_THREAD))
```
I'm trying to launch the window by using DispatcherQueue.TryEnqueue method but that method is not showing in intellisense.
Please guide me that how should I launch the window?
What should I do?

**Comment by akash07k** [comment]
Anyone please?
This is quite urgent for me

**Comment by castorix** [comment]
You can just comment _m_window.Activate();_
Then when I use a Hotkey (with [RegisterHotKey ](https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-registerhotkey))
I can show and activate the window in WM_HOTKEY with _this.Activate();_

**Comment by akash07k** [comment]
@castorix 
Yes, I too have removed the said line from my Application class's OnLaunched method.
Now I'm unable to do m_window.activate when we press the shortcut key.
I'm using H.Hooks lib for keyboard hook.
Can it due to different thread?
Is it possible that keyboard hook is running on different thread?

**Comment by castorix** [comment]
I don't use any external library
I also tested with a Hook ([SetWindowsHookEx](https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-setwindowshookexw)) and it works fine (WH_KEYBOARD_LL and 0 for _dwThreadId_)

**Comment by akash07k** [comment]
So are you able to launch the MainWindow from the raised event by the hook?
@castorix

**Comment by castorix** [comment]
> So are you able to launch the MainWindow from the raised event by the hook? 

Yes, _this.Activate();_ shows the window and sets it to foreground when the key is pressed from the hook procedure

**Comment by akash07k** [comment]
Seems that you are doing it (this.Activate() from the MainWindow class.
And I'm trying it from the App class. This.activate will only be available in MainWindow class not in Application class.
And the thing is that upon my app launches, I don't want to show/launch any window at all.
That's why I'm hooking the keyboard from the application class.
So the thing is that,if I'm doing 
MainWindow window = MainWindow();
window.Activate();
from the OnLaunched method or constructor then it's getting launched.
However if I'm trying the same from the Hook event then it's raising the error.
It seems it is happening because Keyboard  Hook is on another thread.
I think I'm clear now.
If no, then I'll share my exact code for your reference then.
@castorix

**Comment by akash07k** [comment]
Exactly, but if I'm creating the window in key event then it is giving me the error:
The application called an interface that was marshalled for a different thread. (0x8001010E

**Comment by castorix** [comment]
I just tested by creating everything in App.xaml.cs and it still works (but not a good method I think to create a window in a shortcut...)

**Comment by StephenLPeters** [comment]
@marb2000 @bpulliam FYI

**Comment by akash07k** [comment]
@castorix 
I don't know if it is a wrong way. if it is a wrong way then I'd love if someone from MSFT please guides me on this

**Comment by akash07k** [comment]
Not stale

**Comment by xiaopoyang-msft** [comment]
@akash07k Obviously your Keyboard Hook is on another thread, which is indicated by the encountered runtime exception and verified by _H.Hooks_ lib.

> Optimized, runs hooks in a separate thread. ...

Just use [SetWindowsHookEx](https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-setwindowshookexw) as @castorix does.
In consideration of your way, you need to maintain the lifetime of the created window(s) through keystrokes of the hot key.

**Comment by akash07k** [comment]
I actually don't want to use SetWindowsHook because it requires more boilerplate for my usecase.I want to use H.Hooks lib instead. any guidance on that please?
Also,
I'm not clear with:
> In consideration of your way, you need to maintain the lifetime of the created window(s) through keystrokes of the hot key.


> @akash07k Obviously your Keyboard Hook is on another thread, which is indicated by the encountered runtime exception and verified by _H.Hooks_ lib.
> 
> > Optimized, runs hooks in a separate thread. ...
> 
> Just use [SetWindowsHookEx](https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-setwindowshookexw) as @castorix does. In consideration of your way, you need to maintain the lifetime of the created window(s) through keystrokes of the hot key.

**Comment by xiaopoyang-msft** [comment]
`Microsoft.UI.Dispatching.DispatcherQueue dispatcherQueue = Microsoft.UI.Dispatching.DispatcherQueue.GetForCurrentThread();`
You may refer to the `DispatcherQueue` of the UI thread in the _H.Hooks lib_ thread. See [Change CoreDispatcher.RunAsync to DispatcherQueue.TryEnqueue](https://learn.microsoft.com/en-us/windows/apps/windows-app-sdk/migrate-to-windows-app-sdk/guides/threading#change-coredispatcherrunasync-to-dispatcherqueuetryenqueue) and the [WinUI-Gallery](https://github.com/microsoft/WinUI-Gallery/blob/861564349b526edada16b51ff6b78dd105564fa5/WinUIGallery/ControlPages/ClipboardPage.xaml.cs#L33) sample for more information.

> In consideration of your way, you need to maintain the lifetime of the created window(s) through keystrokes of the hot key.

How about a user raising a succession of keystrokes of the hot key? You should avoid creating windows per keystroke.

**Comment by niels9001** [comment]
See the Windows samples in WinUI Gallery for more example code.
